pydm/main_window.py

```diff
@@ -107,166 +105,80 @@ def set_display_widget(self, new_widget):
-            self.handle_open_file_error(filename, e)
-
     def handle_open_file_error(self, filename, error):
         self.statusBar().showMessage("Cannot open file: '{0}', reason: '{1}'...".format(filename, error), 5000)
 
-    # Note: in go(), back(), and forward(), always do history stack manipulation *before* opening the file.
-    # That way, the navigation button enable/disable state will work correctly.  This is stupid, and will be fixed eventually.
-    def go(self, ui_file, macros=None, command_line_args=None):
-        self.forward_stack = []
-        self.open_file(ui_file, macros, command_line_args)
+    def back(self):
+        curr_display = self.display_widget()
+        prev_display = None
+        if curr_display:
+            prev_display = curr_display.previous_display()
 
-    def go_abs(self, ui_file, macros=None, command_line_args=None):
-        self.forward_stack = []
-        self.open_abs_file(filename=ui_file, macros=macros, command_line_args=command_line_args)
+        if not prev_display:
+            logger.error('No display history to execute back navigation.')
+            return
 
-    def back(self):
-        if len(self.back_stack) > 1:
-            if QApplication.keyboardModifiers() == Qt.ShiftModifier:
-                stack_item = self.back_stack[-2]
-                self.new_abs_window(filename=stack_item[0], macros=stack_item[1], command_line_args=stack_item[2])
-            else:
-                self.forward_stack.append(self.back_stack.pop())
-                stack_item = self.back_stack[-1]
-                self.open_abs_file(filename=stack_item[0], macros=stack_item[1], command_line_args=stack_item[2])
+        if QApplication.keyboardModifiers() == Qt.ShiftModifier:
+            load_file(prev_display.loaded_file(),
+                      macros=prev_display.macros,
+                      args=prev_display.args,
+                      target=ScreenTarget.NEW_PROCESS)
+        else:
+            prev_display.set_next_display(curr_display)
+            establish_widget_connections(prev_display)
+            register_widget_rules(prev_display)
+            self.set_display_widget(prev_display)
 
     def forward(self):
-        if len(self.forward_stack) > 0:
-            if QApplication.keyboardModifiers() == Qt.ShiftModifier:
-                stack_item = self.forward_stack[-1]
-                self.new_abs_window(filename=stack_item[0], macros=stack_item[1], command_line_args=stack_item[2])
-            else:
-                stack_item = self.forward_stack.pop()
-                self.open_abs_file(filename=stack_item[0], macros=stack_item[1], command_line_args=stack_item[2])
+        curr_display = self.display_widget()
+        next_display = None
+        if curr_display:
+            next_display = curr_display.next_display()
 
-    def home(self):
-        if self.home_file is None:
+        if not next_display:
+            logger.error('No display history to execute forward navigation.')
```

Comment:
Are the forward/back buttons still getting disabled when there's no history?

Reply:
Addressed via 62acf85